compose/project.py

```diff
@@ -54,7 +54,7 @@ def visit(n):
     return sorted_services
 
 
-class Project(object):
+class Project(ParallelContainerManager):
```

Comment:
My main (only?) problem with the use of inheritance here is that it's hiding the degree of coupling between `ParallelContainerManager` and any class which mixes it in. The contract seems to be:

1. The subclass must implement `containers(self, service_names, stopped)`. It returns a list, whose elements must implement `name` and `is_running`.
2. The subclass must have a `log` attribute, which is an instance of `logging.Logger` (or implements the same interface).

In the second case, I think having a global file-level `log` object removes that coupling. In the first case, it's beginning to feel like `ParallelContainerManager` knows a bit too much about `Project`, or even about containers.

Suppose we had a single method:

```python
def parallel_execute(objects,
                     method,
                     doing_message=None,
                     done_message=None,
                     max_workers=None,
                     args=None,
                     kwargs=None):
    """
    Executes the named method on each object in `objects` in parallel, passing
    along `args` and `kwargs` to each execution.

    Returns `None` once all executions have completed.

    If not `None`, `max_workers` should be an `int` specifying the maximum
    number of concurrent workers to use.

    If not `None`, `doing_message` should be a function which takes a single
    object and returns a string to output once the method has started.

    If not `None`, `done_message` should be a function which takes a single
    object and returns a string to output once the method has finished.
    """
    # ...
```

`project.py` would then look something like:

```python
def stop(self, service_names=None, **options):
    parallel_execute(
        self.containers(service_names),
        'stop',
        doing_message=lambda c: "{}: Stopping...".format(c.name),
        done_message=lambda c: "{}: Stopped".format(c.name),
        kwargs=options,
    )

def kill(self, service_names=None, **options):
    parallel_execute(
        self.containers(service_names),
        'kill',
        doing_message=lambda c: "{}: Killing...".format(c.name),
        done_message=lambda c: "{}: Killed".format(c.name),
        kwargs=options,
    )

def remove(self, service_names=None, **options):
    parallel_execute(
        [c for c in self.containers(service_names) if not c.is_running],
        'remove',
        doing_message=lambda c: "{}: Removing...".format(c.name),
        done_message=lambda c: "{}: Removed".format(c.name),
        kwargs=options,
    )
```

We've taken a duplication hit here, but what I like is that we've got the scary parallel stuff squirrelled away in as small and sealed a box as possible.